src/modules/fundingManager/bondingCurve/formulas/DiscreteCurveMathLib_v1.sol

```diff
@@ -0,0 +1,696 @@
+            uint collateralForPortion_ = _calculateSegmentReserve(
+                initialPrice_,
+                priceIncreasePerStep_,
+                supplyPerStep_,
+                supplyToProcessInSegment_
+            );
+
+            totalReserve_ += collateralForPortion_;
+
+            // Update cumulative supply with actual supply processed
+            cumulativeSupplyProcessed_ += supplyToProcessInSegment_;
+        }
+
+        return totalReserve_;
+    }
+
+    /**
+     * @notice Calculates the amount of issuance tokens a purchaser receives for a given amount of collateral,
+     *         and the actual amount of collateral spent.
+     * @dev Iterates through curve segments starting from the position_ indicated by currentTotalIssuanceSupply_.
+     *      It first determines the starting segment and step. If the starting supply is mid-step,
+     *      it calculates the cost to complete that partial step. Then, it iterates through subsequent
+     *      full steps and segments, consuming the provided collateral.
+     *      The function handles purchases that span multiple steps and segments.
+     *      It ensures that collateral is not overspent and issuance does not exceed curve capacity.
+     *      Collateral calculations for steps are rounded up to favor the protocol.
+     *      Tokens minted for a partial final step (due to budget constraint) are rounded down.
+     *      Reverts if collateralToSpendProvided_ is zero.
+     *      Reverts if segments_ array is empty.
+     *      Caller is responsible for pre-validating the segments_ array structure (e.g., price progression, MAX_SEGMENTS)
+     *      and ensuring currentTotalIssuanceSupply_ does not exceed total curve capacity before calling this function.
+     * @param segments_ Array of PackedSegment configurations for the curve.
+     * @param collateralToSpendProvided_ The amount of collateral the purchaser is providing.
+     * @param currentTotalIssuanceSupply_ The current total issuance supply before this purchase.
+     * @return tokensToMint_ The total amount of issuance tokens to be minted.
+     * @return collateralSpentByPurchaser_ The actual amount of collateral spent from the provided budget.
+     */
+    function _calculatePurchaseReturn(
+        PackedSegment[] memory segments_,
+        uint collateralToSpendProvided_,
+        uint currentTotalIssuanceSupply_
+    )
+        internal
+        pure // Already pure, ensuring it stays
+        returns (uint tokensToMint_, uint collateralSpentByPurchaser_)
+    {
+        if (collateralToSpendProvided_ == 0) {
+            revert
+                IDiscreteCurveMathLib_v1
+                .DiscreteCurveMathLib__ZeroCollateralInput();
+        }
+
+        // Phase 1: Find which segment and step to start purchasing from.
+        uint segmentIndex_ = 0;
+        uint supplyCoveredByPreviousSegments_ = 0;
+
+        if (currentTotalIssuanceSupply_ > 0) {
```

Comment:
these are a lot of potentially very long loops, and running out of gas may be a real problem here if you use a lot of segments.
also, many of the values here are are static (i.e. the segments capacity and total issuance) and could plausibly be cached or caluclated only once on init. In that way, you can do a tree search instead of each time looping over all segments and subsegments.